# <center>Laboratorium Analiza i bazy danych </center>

## <center>Łączenie tabel, podzapytania i funkcje agregujące</center>

Aleksandra Stachniak, grupa 1

## Przykładowe tabele obrazujące łączenie

Do zobrazowania operacji łączenia zostaną użyte tabele:

```sql
CREATE TABLE shape_a (
    id INT PRIMARY KEY,
    shape VARCHAR (100) NOT NULL
);
 
CREATE TABLE shape_b (
    id INT PRIMARY KEY,
    shape VARCHAR (100) NOT NULL
);
```
 
Polecenie CREATE TABLE tworzy tabelę o zadanej nazwie i strukturze. Ogólna postać to:
```sql
CREATE TABLE tab_name (
    col_name1 data_type constrain,
    col_name1 data_type constrain,
    ...
);
```
Należy uzupełnić ją danymi:
```sql
INSERT INTO shape_a (id, shape)
VALUES
    (1, 'Trójkąt'),
    (2, 'Kwadrat'),
    (3, 'Deltoid'),
    (4, 'Traper');
 
INSERT INTO shape_b (id, shape)
VALUES
    (1, 'Kwadrat'),
    (2, 'Trójkąt'),
    (3, 'Romb'),
    (4, 'Równoległobok');
```
Komenda INSERT INTO pozwala na dodanie do tabeli rekordów. Ogólna postać to:

```sql
INSERT INTO tab_name (col1_name, col2_name2, ...) 
VALUES
    (val1_col1, val2_col2),
    (val2_col1, val2_col2),
    ...
```

## Inner join 

Jest to podstawowy rodzaj złączenie. Ten sposób złączenia wybiera  te wiersze, dla których warunek złączenia jest spełniony. W żadnej z łączonych tabel kolumna użyta do łączenia nie może mieć wartości NULL. 

#### Przykład:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
INNER JOIN shape_b b ON a.shape = b.shape;
```
W zapytaniu powyżej użyto *aliasów* nazw tabel i column wynikowych, jest to szczególnie przydatne przy długich nazwach tabel i wprowadza czytelność w zapytaniu.

#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Trójkąt|2|Trójkąt|
|2|Kwadrat|1|Kwadrat|

## OUTER JOIN

Istnieją trzy rodzaje złączeń OUTER:
- LEFT OUTER JOIN,
- RIGHT OUTER JOIN,
- FULL OUTER JOIN.

### LEFT OUTER JOIN

Ten rodzaj złączenie zwróci wszystkie rekordy z lewej tablicy i dopasuje do nich rekordy z prawej tablicy które spełniją zadany warunek złączenia. Jeżeli w prawej tablicy nie występują rekordy spełnijące warunek złączenia z lewą w ich miejscu pojawią się wartości NULL.

#### Przykład 1:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
LEFT JOIN shape_b b ON a.shape = b.shape;
```
#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Trójkąt|2|Trójkąt|
|2|Kwadrat|1|Kwadrat|
|3|Deltoid|NULL|NULL|
|4|Traper|NULL|NULL|

#### Przykład 2:
```sql
SELECT
    b.id id_b,
    b.shape shape_b,
    a.id id_a,
    a.shape shape_a   
FROM
    shape_b b
LEFT JOIN shape_a a ON a.shape = b.shape;
```
#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Kwadrat|2|Kwadrat|
|2|Trójkąt|1|Trójkąt|
|3|Romb|NULL|NULL|
|4|Równoległobok|NULL|NULL|

### RIGHT OUTER JOIN

Działa jak left outer join z tym, że prawa tablica w zapytaniu jest brana w całości.

#### Przykład:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
RIGHT JOIN shape_b b ON a.shape = b.shape;
```

#### Wynik:
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|2|Kwadrat|1|Kwadrat|
|1|Trójkąt|2|Trójkąt|
|NULL|NULL|3|Romb|
|NULL|NULL|4|Równoległobok|


### FULL OUTER JOIN

Jest złączeniem które zwraca:
- wiersze dla których warunek złączenia jest spełniony,
- wiersze z lewej tabeli dla których nie ma odpowiedników w prawej,
- wiersze z prawej tabeli dla których nie ma odpowiedników w lewej. 

#### Przykład:
```sql
SELECT
    a.id id_a,
    a.shape shape_a,
    b.id id_b,
    b.shape shape_b
FROM
    shape_a a
FULL JOIN shape_b b ON a.shape = b.shape;
```
|id_a|shape_a|id_b|shape_b|
|-|-|-|-|
|1|Trójkąt|2|Trójkąt|
|2|Kwadrat|1|Kwadrat|
|3|Deltoid"|NULL|NULL|
|4|Traper|NULL|NULL|
|NULL|NULL|3|Romb|
|NULL|NULL|4|Równoległobok|

## Podzapytania

Podzapytanie zagnieżdżone SELECT znajduje się wewnątrz zewnętrznego zapytania SELECT, np. po klauzuli WHERE, HAVING lub FROM. W przypadku tego rodzaju zapytań w pierwszej kolejności wykonywane są wewnętrzne zapytania SELECT, a ich wynik jest wykorzystywany do zewnętrznego zapytania SELECT. Stąd łatwo zuważyć, że mogą one służyć do poprawy wydajności obsługi zapytania. Należy dobierać podzapytania tak by najbardziej zagnieżdżone podzapytanie zawierało najmniejszy zbiór poszukiwań. 

#### Przykład:
Jeżeli chcemy znaleźć w bazie informację o tytułach filmów zwróconych w zadanym okresie możemy wykonać następujące zapytanie:
```sql
SELECT
   film_id,
   title
FROM
   film
WHERE
   film_id IN (
      SELECT
         inventory.film_id
      FROM
         rental
      INNER JOIN inventory ON inventory.inventory_id = rental.inventory_id
      WHERE
         return_date BETWEEN '2005-05-29'
      AND '2005-05-30'
   );
```

#### Wynik
|film_id|title|
|-|-|
|307|Fellowship Autumn|
|255|Driving Polish|
|388|Gunfight Moon|
|130|Celebrity Horn|
|563|Massacre Usual|
|397|Hanky October|
|...|...|

### Używanie podzapytań

Pod zapytania mogą być używane w :
- SELECT,
- UPDATE,
- DELETE,
- Funkcjach agregujących,
- Do definiowania tabel tymczasowych.

Używając podzapytań zapytania SQL szybko mogą stać się mało czytelne. Przez co będą trudne w zrozumieniu i późniejszym utrzymaniu. W celu analizy zapytań można użyć klauzuli __EXPLAIN__, która przeanalizuje zapytanie. Klauzula ta może służyć również do porównywania wydajności zapytań

#### Przykład:
```sql
EXPLAIN SELECT
   *
FROM
   film
```

## Funkcje agregujące

Funkcje agregujące wykonują obliczenia na zestawie wierszy i zwracają pojedynczy wiersz. PostgreSQL udostępnia wszystkie standardowe funkcje agregujące SQL w następujący sposób:
- AVG () - zwraca średnią wartość.
- COUNT () - zwraca liczbę wartości.
- MAX () - zwraca maksymalną wartość.
- MIN () - zwraca minimalną wartość.
- SUM () - zwraca sumę wszystkich lub różnych wartości.

Pełna lista funkcji agregującej: https://www.postgresql.org/docs/9.5/functions-aggregate.html

Często używamy funkcji agregujących z klauzulą GROUP BY w instrukcji SELECT. W tych przypadkach klauzula GROUP BY dzieli zestaw wyników na grupy wierszy i funkcja agregująca wykonuje obliczenia dla każdej grupy, np. maksimum, minimum, średnia itp. Funkcji agregujących można używać funkcji agregujących jako wyrażeń tylko w następujących klauzulach: SELECT i HAVING.

### GROUP BY
Klauzula GROUP BY dzieli wiersze zwrócone z instrukcji SELECT na grupy. Dla  każdej grupy można zastosować funkcję agregującą, np. SUM aby obliczyć sumę pozycji lub
COUNT aby uzyskać liczbę elementów w grupach.

Poniższa instrukcja ilustruje składnię klauzuli GROUP BY:
```sql
SELECT 
    column_1, 
    aggregate_function(column_2)
FROM 
    tbl_name
GROUP BY 
    column_1;
```
Klauzula GROUP BY musi pojawić się zaraz po klauzuli FROM lub WHERE, n0astępnie GROUP BY zawiera listę  kolumna oddzielonych przecinkami. 

### HAVING
Często używamy klauzuli HAVING w połączeniu z klauzulą GROUP BY do filtrowania wierszy grup
które nie spełniają określonego warunku.

Poniższa instrukcja ilustruje typową składnię klauzuli HAVING:
```sql
SELECT
    column_1,
    aggregate_function (column_2)
FROM
    tbl_name
GROUP BY
    column_1
HAVING
    condition;
```
Klauzula HAVING ustawia warunek dla wierszy grup utworzonych przez klauzulę GROUP BY.  

Klauzula GROUP BY ma zastosowanie, podczas gdy klauzula WHERE określa wcześniej warunki dla poszczególnych wierszy.

## Zadania wprowadzające
Wykonaj zapytania przy użyciu DBMS:  
  
1. Znajdź listę wszystkich filmów o tej samej długości.
2. Znajdź wszystkich klientów mieszkających w tym samym mieście.
3. Oblicz średni koszt wypożyczenia wszystkich filmów.
4. Oblicz i wyświetl liczbę filmów we wszystkich kategoriach.
5. Wyświetl liczbę wszystkich klientów pogrupowanych według kraju.
6. Wyświetl informacje o sklepie, który ma więcej niż 100 klientów i mniej niż 300 klientów.
7. Wybierz wszystkich klientów, którzy oglądali filmy ponad 200 godzin.
8. Oblicz średnią wartość wypożyczenia filmu.
9. Oblicz średnią wartość długości filmu we wszystkich kategoriach.
10. Znajdź najdłuższe tytuły filmowe we wszystkich kategoriach.
11. Znajdź najdłuższy film we wszystkich kategoriach. Porównaj wynik z pkt 10.

**Łączenie się za pomocą psycopg2 i pandas**

In [1]:
import psycopg2 as pg
import pandas as pd

connection = pg.connect(host='pgsql-196447.vipserv.org', port=5432, dbname='wbauer_adb', user='wbauer_adb', password='adb2020')

**1. Znajdź listę wszystkich filmów o tej samej długości.**

In [2]:
df_1 = pd.read_sql('select count(length) count, length from film group by length', con=connection)
df_1

count  length
0        4     116
1        8     184
2        6      87
3        7      71
4        5      68
..     ...     ...
135      7     148
136      7     129
137      6     130
138      7     143
139      7      58

[140 rows x 2 columns]

**2. Znajdź wszystkich klientów mieszkających w tym samym mieście.**

In [3]:
df_2 = pd.read_sql(("select city.city, count(customer_id) from customer " + 
                 "inner join address on customer.address_id = address.address_id " +
                 "inner join city on address.city_id = city.city_id " + 
                 "group by city.city_id"), con=connection)
df_2 

city  count
0         Gorontalo      1
1        Botshabelo      1
2            Siegen      1
3        Klerksdorp      1
4           Udaipur      1
..              ...    ...
592  San Bernardino      1
593     Tallahassee      1
594           Beira      1
595       Barcelona      1
596        Duisburg      1

[597 rows x 2 columns]

**3. Oblicz średni koszt wypożyczenia wszystkich filmów.**

In [4]:
df_3 = pd.read_sql("select avg(amount) from payment", con=connection)
df_3

avg
0  4.200606

**Odpowiedź:** Interesuje nas średni koszt wypożyczenia wszystkich filmów, zatem tej informacji poszukuję w tabeli z danymi dotyczącymi płatności. Koszt ten wynosi 4.200606.

**4. Oblicz i wyświetl liczbę filmów we wszystkich kategoriach.**

In [5]:
df4 = pd.read_sql("select name, count(film_category.film_id) from category " + 
                 "inner join film_category on category.category_id = film_category.category_id " + 
                 "group by name", con=connection)
df4

name  count
0        Sports     74
1      Classics     57
2           New     63
3        Family     69
4        Comedy     58
5     Animation     66
6        Travel     57
7         Music     51
8         Drama     62
9        Horror     56
10       Sci-Fi     61
11        Games     61
12  Documentary     68
13      Foreign     73
14       Action     64
15     Children     60

**5. Wyświetl liczbę wszystkich klientów pogrupowanych według kraju.**

In [6]:
df_5 = pd.read_sql(("select country, count(customer_id) from customer " + 
                 "inner join address on customer.address_id = address.address_id " +
                 "inner join city on address.city_id = city.city_id " + 
                 "inner join country on city.country_id = country.country_id " + 
                 "group by country"), con=connection)
df_5

country  count
0              Bangladesh      3
1               Indonesia     14
2               Venezuela      7
3                Cameroon      2
4          Czech Republic      1
..                    ...    ...
103              Tanzania      3
104                Poland      8
105             Greenland      1
106         French Guiana      1
107  Virgin Islands, U.S.      1

[108 rows x 2 columns]

**6. Wyświetl informacje o sklepie, który ma więcej niż 100 klientów i mniej niż 300 klientów.**

In [7]:
df_6 = pd.read_sql("""select count(customer.customer_id), store.store_id, store.manager_staff_id, store.address_id from store
    inner join staff on staff.store_id = store.store_id
    inner join payment on payment.staff_id = staff.staff_id
    inner join customer on customer.customer_id = payment.customer_id
    group by store.store_id having (count(customer.customer_id) > 100 and count(customer.customer_id) < 300)""", con=connection)
df_6

Empty DataFrame
Columns: [count, store_id, manager_staff_id, address_id]
Index: []

**Odpowiedź:** Jak widać powyżej nie ma sklepów spełniających oba warunki. Mogę jedynie wyświetlić informacje o sklepach, które mają więcej niż 100 klientów, bez górnego ograniczenia, co uczyniłam poniżej.

In [8]:
df_6 = pd.read_sql("""select count(customer.customer_id), store.store_id, store.manager_staff_id, store.address_id from store
    inner join staff on staff.store_id = store.store_id
    inner join payment on payment.staff_id = staff.staff_id
    inner join customer on customer.customer_id = payment.customer_id
    group by store.store_id having (count(customer.customer_id) > 100)""", con=connection)
df_6

count  store_id  manager_staff_id  address_id
0   7304         2                 2           2
1   7292         1                 1           1

**7. Wybierz wszystkich klientów, którzy oglądali filmy ponad 200 godzin.**

In [9]:
df_7 = pd.read_sql("""select customer.customer_id, sum(film.length) from customer 
    inner join rental on customer.customer_id = rental.customer_id
    inner join inventory on rental.inventory_id = inventory.inventory_id
    inner join film on inventory.film_id = film.film_id
    group by customer.customer_id having sum(film.length) > 200""", con=connection)
df_7

customer_id   sum
0             87  3446
1            184  2649
2            477  2750
3            273  3606
4            550  3309
..           ...   ...
594          449  2119
595          520  3736
596           64  3515
597           55  2756
598          148  5360

[599 rows x 2 columns]

**8. Oblicz średnią wartość wypożyczenia filmu.**

In [10]:
df_8 = pd.read_sql("select avg(rental_rate) from film", con=connection)
df_8

avg
0  2.98

**Odpowiedź:** W odróżnieniu od polecenia w punkcie 3 tutaj interesuje nas średnia wartość wypożeczenia filmu, zatem tej informacji poszukuję w tabeli z danymi dotyczącymi wypożyczeń. Wartość ta jest równa 2.98.

**9. Oblicz średnią wartość długości filmu we wszystkich kategoriach.**

In [11]:
df_9 = pd.read_sql(("""select name, avg(length) from category
    inner join film_category on category.category_id = film_category.category_id
    inner join film on film.film_id = film_category.film_id
    group by name"""), con=connection)
df_9 

name         avg
0        Sports  128.202703
1      Classics  111.666667
2           New  111.126984
3        Family  114.782609
4        Comedy  115.827586
5     Animation  111.015152
6        Travel  113.315789
7         Music  113.647059
8         Drama  120.838710
9        Horror  112.482143
10       Sci-Fi  108.196721
11        Games  127.836066
12  Documentary  108.750000
13      Foreign  121.698630
14       Action  111.609375
15     Children  109.800000

**10. Znajdź najdłuższe tytuły filmowe we wszystkich kategoriach.**

In [12]:
df_10 = pd.read_sql("""select name, title, length from category
    inner join film_category on category.category_id = film_category.category_id
    inner join film on film.film_id = film_category.film_id
    where (name, length) in (select name, max(length) from category
        inner join film_category on category.category_id = film_category.category_id
        inner join film on film.film_id = film_category.film_id
        group by name)""", con=connection)
df_10

name               title  length
0        Horror    Analyze Hoosiers     181
1         Games       Chicago North     185
2      Classics   Conspiracy Spirit     184
3        Comedy      Control Anthem     185
4       Foreign    Crystal Breaking     184
5        Action      Darn Forrester     185
6           New      Frontier Cabin     183
7      Children         Fury Murder     178
8     Animation         Gangs Pride     185
9         Music           Home Pity     185
10        Drama       Jacket Frisco     181
11       Family      King Evolution     184
12       Horror       Love Suicides     181
13       Travel       Muscle Bright     185
14    Animation        Pond Seattle     185
15       Sports     Smoochy Control     184
16       Sci-Fi  Soldiers Evolution     185
17      Foreign      Sorority Queen     184
18       Travel   Sweet Brotherhood     185
19  Documentary           Wife Turn     183
20       Action        Worst Banger     185
21     Children      Wrong Behavior     178
22  Documentary      Young Language     183

**11. Znajdź najdłuższy film we wszystkich kategoriach. Porównaj wynik z pkt 10.**

In [13]:
df_11 = pd.read_sql("""select title, length from film 
    where length in (select max(length) from film)""", con=connection)
df_11

title  length
0       Chicago North     185
1      Control Anthem     185
2      Darn Forrester     185
3         Gangs Pride     185
4           Home Pity     185
5       Muscle Bright     185
6        Pond Seattle     185
7  Soldiers Evolution     185
8   Sweet Brotherhood     185
9        Worst Banger     185

## Zadanie implementacyjne
Zaimplementuj wszystkie funkcje w pliku main.py zgodnie z opisem a następnie przetestuj je w notatniku.

In [14]:
import main

In [15]:
df = main.film_in_category(1)
df

title               languge category
0             Amadeus Holy  English                Action
1          American Circus  English                Action
2       Antitrust Tomatoes  English                Action
3            Ark Ridgemont  English                Action
4      Barefoot Manchurian  English                Action
..                     ...                   ...      ...
59         Uprising Uptown  English                Action
60  Waterfront Deliverance  English                Action
61           Werewolf Lola  English                Action
62            Women Dorado  English                Action
63            Worst Banger  English                Action

[64 rows x 3 columns]

In [16]:
df = main.number_films_in_category(1)
df

category  count
0   Action     64

In [17]:
df = main.number_film_by_length(0, 46)
df

length  count
0      46      5

In [18]:
df = main.client_from_city("Athenai")
df

city first_name last_name
0  Athenai      Linda  Williams

In [19]:
df = main.avg_amount_by_length(48)
df

length       avg
0      48  4.295389

In [20]:
df = main.client_by_sum_length(1200)
df

first_name last_name   sum
0        Brian     Wyman  1265
1      Antonio      Meek  1451
2        Leona    Obrien  1588
3    Katherine    Rivera  1615
4      Tiffany    Jordan  1667
..         ...       ...   ...
594     Wesley      Bull  4808
595      Clara      Shaw  4808
596      Tammy   Sanders  5065
597    Eleanor      Hunt  5360
598       Karl      Seal  5388

[599 rows x 3 columns]

In [21]:
df = main.category_statistic_length("Action")
df

category         avg   sum  min  max
0   Action  111.609375  7143   47  185